File: runtime/microvm/session.go
package microvm

import (
	"strings"
	"time"
)

const (
	// DefaultSessionProviderID is the provider id used by deterministic AppTheory registry clients.
	DefaultSessionProviderID = "apptheory.microvm.registry"
	// AWSLambdaMicroVMProviderID is the provider id for the official AWS Lambda MicroVM provider.
	AWSLambdaMicroVMProviderID = "aws.lambda.microvm"
)

// SessionKey identifies a MicroVM session by tenant, namespace, and session_id.
type SessionKey struct {
	TenantID  string `json:"tenant_id"`
	Namespace string `json:"namespace"`
	SessionID string `json:"session_id"`
}

// SessionTokenMetadata is the only token information a MicroVM registry record may persist.
//
// It deliberately contains no plaintext token, bearer token, X-aws-proxy-auth header value,
// provider credential, raw SDK object, or provider pagination token.
type SessionTokenMetadata struct {
	TokenID   string    `json:"token_id"`
	TokenType string    `json:"token_type"`
	ExpiresAt time.Time `json:"expires_at"`
	Scope     []string  `json:"scope"`
}

// CreateSessionInput is the safe client input for creating a session.
type CreateSessionInput struct {
	RequestID           string      `json:"request_id"`
	TenantID            string      `json:"tenant_id"`
	Namespace           string      `json:"namespace"`
	SessionID           string      `json:"session_id"`
	ImageRef            string      `json:"image_ref"`
	NetworkConnectorRef string      `json:"network_connector_ref"`
	SessionSpec         SessionSpec `json:"session_spec"`
	ControllerID        string      `json:"controller_id"`
	AuthSubject         string      `json:"auth_subject"`
	Now                 time.Time   `json:"now"`
}

// SessionCommandInput is the safe client input for start/stop commands.
type SessionCommandInput struct {
	RequestID    string         `json:"request_id"`
	TenantID     string         `json:"tenant_id"`
	Namespace    string         `json:"namespace"`
	SessionID    string         `json:"session_id"`
	ControllerID string         `json:"controller_id"`
	AuthSubject  string         `json:"auth_subject"`
	DesiredState LifecycleState `json:"desired_state"`
	Now          time.Time      `json:"now"`
}

// SessionQueryInput is the safe client input for status/session queries.
type SessionQueryInput struct {
	RequestID   string `json:"request_id"`
	TenantID    string `json:"tenant_id"`
	Namespace   string `json:"namespace"`
	SessionID   string `json:"session_id"`
	AuthSubject string `json:"auth_subject"`
}

// SessionRecord is the safe durable-session shape AppTheory exposes to clients and registries.
type SessionRecord struct {
	TenantID                    string                 `json:"tenant_id"`
	Namespace                   string                 `json:"namespace"`
	SessionID                   string                 `json:"session_id"`
	State                       LifecycleState         `json:"state"`
	DesiredState                LifecycleState         `json:"desired_state"`
	Endpoint                    string                 `json:"endpoint,omitempty"`
	MicroVMID                   string                 `json:"microvm_id,omitempty"`
	ProviderID                  string                 `json:"provider_id"`
	ProviderMicroVMID           string                 `json:"provider_microvm_id,omitempty"`
	ProviderState               string                 `json:"provider_state"`
	AWSLifecycleState           string                 `json:"aws_lifecycle_state"`
	ImageRef                    string                 `json:"image_ref"`
	ImageVersion                string                 `json:"image_version,omitempty"`
	NetworkConnectorRef         string                 `json:"network_connector_ref"`
	IngressNetworkConnectorRefs []string               `json:"ingress_network_connector_refs,omitempty"`
	EgressNetworkConnectorRefs  []string               `json:"egress_network_connector_refs,omitempty"`
	ControllerID                string                 `json:"controller_id"`
	CreatedAt                   time.Time              `json:"created_at"`
	UpdatedAt                   time.Time              `json:"updated_at"`
	LastObservedAt              time.Time              `json:"last_observed_at"`
	ProviderStartedAt           time.Time              `json:"provider_started_at,omitempty"`
	ProviderTerminatedAt        time.Time              `json:"provider_terminated_at,omitempty"`
	ExpiresAt                   time.Time              `json:"expires_at"`
	Generation                  int64                  `json:"generation"`
	LastAction                  Command                `json:"last_action"`
	LastCommandID               string                 `json:"last_command_id"`
	AuthSubject                 string                 `json:"auth_subject"`
	ReasonMetadata              map[string]string      `json:"reason_metadata,omitempty"`
	StatusMetadata              map[string]string      `json:"status_metadata,omitempty"`
	TokenMetadata               []SessionTokenMetadata `json:"token_metadata,omitempty"`
	Metadata                    map[string]string      `json:"metadata,omitempty"`
}

// SessionStatus is the safe controller status response from a constrained client.
type SessionStatus struct {
	TenantID        string         `json:"tenant_id"`
	Namespace       string         `json:"namespace"`
	SessionID       string         `json:"session_id"`
	State           LifecycleState `json:"state"`
	DesiredState    LifecycleState `json:"desired_state"`
	LifecycleState  LifecycleState `json:"lifecycle_state"`
	Endpoint        string         `json:"endpoint,omitempty"`
	MicroVMID       string         `json:"microvm_id,omitempty"`
	LastAction      Command        `json:"last_action"`
	LastTransition  time.Time      `json:"last_transition"`
	RegistryVersion int64          `json:"registry_version"`
}

// ValidateSessionRecord fails closed when a session record is incomplete or contains forbidden metadata.
func ValidateSessionRecord(record SessionRecord) error {
	record = normalizeSessionRecord(record)
	if err := validateSessionRecordIdentity(record); err != nil {
		return err
	}
	if err := validateSessionRecordRegistryFields(record); err != nil {
		return err
	}
	if err := validateSessionRecordProviderFields(record); err != nil {
		return err
	}
	if !validLifecycleState(record.State) || !validLifecycleState(record.DesiredState) {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session record state is unsupported", record.LastCommandID)
	}
	if err := validateSafeMetadata(record.Metadata, record.LastCommandID); err != nil {
		return err
	}
	if err := validateSafeMetadata(record.ReasonMetadata, record.LastCommandID); err != nil {
		return err
	}
	return validateSafeMetadata(record.StatusMetadata, record.LastCommandID)
}

func validateSessionRecordIdentity(record SessionRecord) error {
	if record.TenantID == "" || record.Namespace == "" || record.SessionID == "" || record.State == "" || record.DesiredState == "" {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session record is incomplete", record.LastCommandID)
	}
	if record.ImageRef == "" || record.NetworkConnectorRef == "" || record.ControllerID == "" || record.LastCommandID == "" || record.AuthSubject == "" {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session record is incomplete", record.LastCommandID)
	}
	if !validCommand(record.LastAction) {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session record last action is unsupported", record.LastCommandID)
	}
	return nil
}

func validateSessionRecordRegistryFields(record SessionRecord) error {
	if record.CreatedAt.IsZero() || record.UpdatedAt.IsZero() || record.ExpiresAt.IsZero() || record.LastObservedAt.IsZero() || record.Generation <= 0 {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session record registry fields are incomplete", record.LastCommandID)
	}
	return nil
}

func validateSessionRecordProviderFields(record SessionRecord) error {
	if record.ProviderID == "" || record.ProviderState == "" || record.AWSLifecycleState == "" {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session record provider fields are incomplete", record.LastCommandID)
	}
	for _, field := range append([]string{
		record.Endpoint,
		record.MicroVMID,
		record.ProviderID,
		record.ProviderMicroVMID,
		record.ProviderState,
		record.AWSLifecycleState,
		record.ImageRef,
		record.ImageVersion,
		record.NetworkConnectorRef,
	}, append(append([]string{}, record.IngressNetworkConnectorRefs...), record.EgressNetworkConnectorRefs...)...) {
		if err := validateSafeFieldValue(field, record.LastCommandID); err != nil {
			return err
		}
	}
	for _, token := range record.TokenMetadata {
		if err := ValidateSessionTokenMetadata(token, record.LastCommandID); err != nil {
			return err
		}
	}
	return nil
}

// ValidateSessionTokenMetadata fails closed if token metadata is incomplete or carries token plaintext.
func ValidateSessionTokenMetadata(token SessionTokenMetadata, requestID string) error {
	token = normalizeSessionTokenMetadata(token)
	if token.TokenID == "" || token.TokenType == "" || token.ExpiresAt.IsZero() || len(token.Scope) == 0 {
		return safeError(ErrorCodeTokenSafetyViolation, "apptheory: microvm session token metadata is incomplete", requestID)
	}
	if err := validateSafeFieldValue(token.TokenID, requestID); err != nil {
		return err
	}
	if err := validateSafeFieldValue(token.TokenType, requestID); err != nil {
		return err
	}
	for _, scope := range token.Scope {
		if err := validateSafeFieldValue(scope, requestID); err != nil {
			return err
		}
	}
	return nil
}

// SessionTokenMetadataFromProviderToken converts sanitized provider token output into registry-safe metadata.
func SessionTokenMetadataFromProviderToken(token ProviderToken) (SessionTokenMetadata, error) {
	token = normalizeProviderToken(token)
	if err := ValidateProviderToken(token); err != nil {
		return SessionTokenMetadata{}, err
	}
	metadata := SessionTokenMetadata{
		TokenID:   token.TokenID,
		TokenType: token.TokenType,
		ExpiresAt: token.ExpiresAt,
		Scope:     append([]string(nil), token.Scope...),
	}
	if err := ValidateSessionTokenMetadata(metadata, ""); err != nil {
		return SessionTokenMetadata{}, err
	}
	return metadata, nil
}

// ValidateSessionStatus fails closed when a status response is incomplete.
func ValidateSessionStatus(status SessionStatus) error {
	status = normalizeSessionStatus(status)
	if status.TenantID == "" || status.Namespace == "" || status.SessionID == "" || status.State == "" || status.DesiredState == "" || status.LifecycleState == "" || status.LastAction == "" || status.LastTransition.IsZero() || status.RegistryVersion <= 0 {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session status is incomplete", "")
	}
	if !validLifecycleState(status.State) || !validLifecycleState(status.DesiredState) || !validLifecycleState(status.LifecycleState) {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session status state is unsupported", "")
	}
	if !validCommand(status.LastAction) {
		return safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm session status last action is unsupported", "")
	}
	return nil
}

// Key returns the tenant/namespace/session key for a session record.
func (r SessionRecord) Key() SessionKey {
	return SessionKey{TenantID: strings.TrimSpace(r.TenantID), Namespace: strings.TrimSpace(r.Namespace), SessionID: strings.TrimSpace(r.SessionID)}
}

// Key returns the tenant/namespace/session key for a status record.
func (s SessionStatus) Key() SessionKey {
	return SessionKey{TenantID: strings.TrimSpace(s.TenantID), Namespace: strings.TrimSpace(s.Namespace), SessionID: strings.TrimSpace(s.SessionID)}
}

func normalizeSessionRecord(record SessionRecord) SessionRecord {
	record.TenantID = strings.TrimSpace(record.TenantID)
	record.Namespace = strings.TrimSpace(record.Namespace)
	record.SessionID = strings.TrimSpace(record.SessionID)
	record.State = LifecycleState(strings.TrimSpace(string(record.State)))
	record.DesiredState = LifecycleState(strings.TrimSpace(string(record.DesiredState)))
	record.Endpoint = strings.TrimSpace(record.Endpoint)
	record.MicroVMID = strings.TrimSpace(record.MicroVMID)
	record.ProviderID = strings.TrimSpace(record.ProviderID)
	record.ProviderMicroVMID = strings.TrimSpace(record.ProviderMicroVMID)
	record.ProviderState = strings.TrimSpace(record.ProviderState)
	record.AWSLifecycleState = strings.TrimSpace(record.AWSLifecycleState)
	record.ImageRef = strings.TrimSpace(record.ImageRef)
	record.ImageVersion = strings.TrimSpace(record.ImageVersion)
	record.NetworkConnectorRef = strings.TrimSpace(record.NetworkConnectorRef)
	record.IngressNetworkConnectorRefs = normalizeStringSlice(record.IngressNetworkConnectorRefs)
	record.EgressNetworkConnectorRefs = normalizeStringSlice(record.EgressNetworkConnectorRefs)
	record.ControllerID = strings.TrimSpace(record.ControllerID)
	record.CreatedAt = record.CreatedAt.UTC()
	record.UpdatedAt = record.UpdatedAt.UTC()
	record.LastObservedAt = record.LastObservedAt.UTC()
	record.ProviderStartedAt = record.ProviderStartedAt.UTC()
	record.ProviderTerminatedAt = record.ProviderTerminatedAt.UTC()
	record.ExpiresAt = record.ExpiresAt.UTC()
	record.LastAction = normalizeCommand(record.LastAction)
	record.LastCommandID = strings.TrimSpace(record.LastCommandID)
	record.AuthSubject = strings.TrimSpace(record.AuthSubject)
	record.ReasonMetadata = cloneStringMap(record.ReasonMetadata)
	record.StatusMetadata = cloneStringMap(record.StatusMetadata)
	record.TokenMetadata = cloneSessionTokenMetadata(record.TokenMetadata)
	record.Metadata = cloneStringMap(record.Metadata)
	return record
}

func normalizeSessionStatus(status SessionStatus) SessionStatus {
	status.TenantID = strings.TrimSpace(status.TenantID)
	status.Namespace = strings.TrimSpace(status.Namespace)
	status.SessionID = strings.TrimSpace(status.SessionID)
	status.State = LifecycleState(strings.TrimSpace(string(status.State)))
	status.DesiredState = LifecycleState(strings.TrimSpace(string(status.DesiredState)))
	status.LifecycleState = LifecycleState(strings.TrimSpace(string(status.LifecycleState)))
	status.Endpoint = strings.TrimSpace(status.Endpoint)
	status.MicroVMID = strings.TrimSpace(status.MicroVMID)
	status.LastAction = normalizeCommand(status.LastAction)
	return status
}

func validLifecycleState(state LifecycleState) bool {
	state = LifecycleState(strings.TrimSpace(string(state)))
	for _, valid := range requiredLifecycleStates() {
		if state == valid {
			return true
		}
	}
	return false
}

func validCommand(command Command) bool {
	command = normalizeCommand(command)
	switch command {
	case CommandCreate, CommandStart, CommandStop, CommandStatus, CommandSession:
		return true
	default:
		return false
	}
}

func normalizeSessionTokenMetadata(token SessionTokenMetadata) SessionTokenMetadata {
	token.TokenID = strings.TrimSpace(token.TokenID)
	token.TokenType = strings.TrimSpace(token.TokenType)
	token.ExpiresAt = token.ExpiresAt.UTC()
	token.Scope = normalizeStringSlice(token.Scope)
	return token
}

func cloneSessionTokenMetadata(tokens []SessionTokenMetadata) []SessionTokenMetadata {
	if len(tokens) == 0 {
		return nil
	}
	out := make([]SessionTokenMetadata, 0, len(tokens))
	for _, token := range tokens {
		normalized := normalizeSessionTokenMetadata(token)
		if normalized.TokenID != "" || normalized.TokenType != "" || !normalized.ExpiresAt.IsZero() || len(normalized.Scope) > 0 {
			out = append(out, normalized)
		}
	}
	return out
}

File: py/src/apptheory/microvm/session.py
from __future__ import annotations

import hashlib
import json as jsonlib
import os
import time
from collections.abc import Callable, Iterable
from dataclasses import dataclass, field
from typing import Any, Literal, cast

# ruff: noqa: F401,F405
from .foundation import *  # noqa: F403
from .model import *  # noqa: F403
from .shared import *  # noqa: F403


def validate_microvm_session_record(record: MicroVMSessionRecord) -> None:
    normalized = _normalize_session_record(record)
    if (
        not normalized.tenant_id
        or not normalized.namespace
        or not normalized.session_id
        or not normalized.state
        or not normalized.desired_state
        or not normalized.provider_id
        or not normalized.provider_state
        or not normalized.aws_lifecycle_state
        or not normalized.image_ref
        or not normalized.network_connector_ref
        or not normalized.controller_id
        or not normalized.last_action
        or not normalized.last_command_id
        or not normalized.auth_subject
    ):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session record is incomplete",
            normalized.last_command_id,
        )
    if (
        normalized.created_at <= 0
        or normalized.updated_at <= 0
        or normalized.last_observed_at <= 0
        or normalized.expires_at <= 0
        or normalized.generation <= 0
    ):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session record registry fields are incomplete",
            normalized.last_command_id,
        )
    if not _valid_microvm_command(normalized.last_action):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session record last action is unsupported",
            normalized.last_command_id,
        )
    if not _valid_lifecycle_state(normalized.state) or not _valid_lifecycle_state(normalized.desired_state):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session record state is unsupported",
            normalized.last_command_id,
        )
    provider_err = _validate_session_provider_fields(normalized)
    if provider_err:
        raise provider_err
    metadata_err = _validate_safe_metadata(normalized.metadata, normalized.last_command_id)
    if metadata_err:
        raise metadata_err
    reason_err = _validate_safe_metadata(normalized.reason_metadata, normalized.last_command_id)
    if reason_err:
        raise reason_err
    status_err = _validate_safe_metadata(normalized.status_metadata, normalized.last_command_id)
    if status_err:
        raise status_err
    reason_err = _validate_safe_metadata(normalized.reason_metadata, normalized.last_command_id)
    if reason_err:
        raise reason_err
    status_err = _validate_safe_metadata(normalized.status_metadata, normalized.last_command_id)
    if status_err:
        raise status_err


def _validate_session_provider_fields(record: MicroVMSessionRecord) -> MicroVMSafeError | None:
    fields = [
        record.endpoint,
        record.microvm_id,
        record.provider_id,
        record.provider_microvm_id,
        record.provider_state,
        record.aws_lifecycle_state,
        record.image_ref,
        record.image_version,
        record.network_connector_ref,
        *record.ingress_network_connector_refs,
        *record.egress_network_connector_refs,
    ]
    for item in fields:
        err = _validate_safe_field_value(item, record.last_command_id)
        if err:
            return err
    for token in record.token_metadata:
        try:
            validate_microvm_session_token_metadata(token, record.last_command_id)
        except MicroVMSafeError as exc:
            return exc
    return None


def validate_microvm_session_token_metadata(
    token: MicroVMSessionTokenMetadata | dict[str, Any], request_id: str = ""
) -> None:
    normalized = _normalize_session_token_metadata(token)
    if not normalized.token_id or not normalized.token_type or normalized.expires_at <= 0 or not normalized.scope:
        raise _safe_error(
            MICROVM_ERROR_TOKEN_SAFETY_VIOLATION,
            "apptheory: microvm session token metadata is incomplete",
            request_id,
        )
    for item in [normalized.token_id, normalized.token_type, *normalized.scope]:
        err = _validate_safe_field_value(item, request_id)
        if err:
            raise err


def microvm_session_token_metadata_from_provider_token(
    token: MicroVMProviderToken | dict[str, Any],
) -> MicroVMSessionTokenMetadata:
    normalized = _normalize_provider_token(token)
    validate_microvm_provider_token(normalized)
    metadata = MicroVMSessionTokenMetadata(
        token_id=normalized.token_id,
        token_type=normalized.token_type,
        expires_at=normalized.expires_at,
        scope=list(normalized.scope),
    )
    validate_microvm_session_token_metadata(metadata)
    return metadata


def validate_microvm_session_status(status: MicroVMSessionStatus) -> None:
    normalized = _normalize_session_status(status)
    if (
        not normalized.tenant_id
        or not normalized.namespace
        or not normalized.session_id
        or not normalized.state
        or not normalized.desired_state
        or not normalized.lifecycle_state
        or not normalized.last_action
        or normalized.last_transition <= 0
        or normalized.registry_version <= 0
    ):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session status is incomplete",
            "",
        )
    if (
        not _valid_lifecycle_state(normalized.state)
        or not _valid_lifecycle_state(normalized.desired_state)
        or not _valid_lifecycle_state(normalized.lifecycle_state)
    ):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session status state is unsupported",
            "",
        )
    if not _valid_microvm_command(normalized.last_action):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session status last action is unsupported",
            "",
        )


def microvm_session_key(record: MicroVMSessionRecord) -> tuple[str, str, str]:
    return (str(record.tenant_id).strip(), str(record.namespace).strip(), str(record.session_id).strip())


def microvm_session_registry_table_name() -> str:
    return (
        str(os.environ.get(MICROVM_SESSION_REGISTRY_TABLE_ENV, "") or "").strip() or MICROVM_SESSION_REGISTRY_TABLE_NAME
    )


def microvm_session_registry_partition_key(tenant_id: str, namespace: str) -> str:
    tenant = str(tenant_id or "").strip()
    ns = str(namespace or "").strip()
    if not tenant or not ns:
        return ""
    return f"TENANT#{tenant}#NAMESPACE#{ns}"


def microvm_session_registry_sort_key(session_id: str) -> str:
    session = str(session_id or "").strip()
    if not session:
        return ""
    return f"SESSION#{session}"


def microvm_session_registry_model_definition(*, table_name: str | None = None) -> Any:
    try:
        from theorydb_py import ModelDefinition  # type: ignore[import-not-found]

        return ModelDefinition.from_dataclass(
            MicroVMSessionRegistryRecord,
            table_name=str(table_name or "").strip() or None,
        )
    except Exception as exc:
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session registry requires TableTheory model support",
            "",
        ) from exc


def validate_microvm_session_registry_record(record: MicroVMSessionRegistryRecord | dict[str, Any]) -> None:
    normalized = _normalize_session_registry_record(record)
    validate_microvm_session_record(_session_record_from_registry_no_validate(normalized))
    if not normalized.pk or not normalized.sk or normalized.ttl <= 0 or normalized.version <= 0:
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session registry keys are incomplete",
            normalized.last_command_id,
        )
    if normalized.pk != microvm_session_registry_partition_key(
        normalized.tenant_id, normalized.namespace
    ) or normalized.sk != microvm_session_registry_sort_key(normalized.session_id):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session registry tenant/session key mismatch",
            normalized.last_command_id,
        )
    if normalized.ttl != int(normalized.expires_at):
        raise _safe_error(
            MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
            "apptheory: microvm session registry ttl mismatch",
            normalized.last_command_id,
        )
    metadata_err = _validate_safe_metadata(normalized.metadata, normalized.last_command_id)
    if metadata_err:
        raise metadata_err


def microvm_session_record_to_registry_record(record: MicroVMSessionRecord) -> MicroVMSessionRegistryRecord:
    normalized = _normalize_session_record(record)
    validate_microvm_session_record(normalized)
    registry = MicroVMSessionRegistryRecord(
        pk=microvm_session_registry_partition_key(normalized.tenant_id, normalized.namespace),
        sk=microvm_session_registry_sort_key(normalized.session_id),
        tenant_id=normalized.tenant_id,
        namespace=normalized.namespace,
        session_id=normalized.session_id,
        state=normalized.state,
        desired_state=normalized.desired_state,
        endpoint=normalized.endpoint,
        microvm_id=normalized.microvm_id,
        provider_id=normalized.provider_id,
        provider_microvm_id=normalized.provider_microvm_id,
        provider_state=normalized.provider_state,
        aws_lifecycle_state=normalized.aws_lifecycle_state,
        image_ref=normalized.image_ref,
        image_version=normalized.image_version,
        network_connector_ref=normalized.network_connector_ref,
        ingress_network_connector_refs=list(normalized.ingress_network_connector_refs),
        egress_network_connector_refs=list(normalized.egress_network_connector_refs),
        controller_id=normalized.controller_id,
        created_at=normalized.created_at,
        updated_at=normalized.updated_at,
        last_observed_at=normalized.last_observed_at,
        provider_started_at=normalized.provider_started_at,
        provider_terminated_at=normalized.provider_terminated_at,
        expires_at=normalized.expires_at,
        ttl=int(normalized.expires_at),
        generation=normalized.generation,
        version=normalized.generation,
        last_action=normalized.last_action,
        last_command_id=normalized.last_command_id,
        auth_subject=normalized.auth_subject,
        reason_metadata=_clone_string_map(normalized.reason_metadata),
        status_metadata=_clone_string_map(normalized.status_metadata),
        token_metadata=_clone_session_token_metadata_list(normalized.token_metadata),
        metadata=_clone_string_map(normalized.metadata),
    )
    validate_microvm_session_registry_record(registry)
    return registry


def microvm_session_from_registry_record(record: MicroVMSessionRegistryRecord | dict[str, Any]) -> MicroVMSessionRecord:
    normalized = _normalize_session_registry_record(record)
    validate_microvm_session_registry_record(normalized)
    out = _session_record_from_registry_no_validate(normalized)
    validate_microvm_session_record(out)
    return out


__all__ = [
    "_validate_session_provider_fields",
    "microvm_session_from_registry_record",
    "microvm_session_key",
    "microvm_session_record_to_registry_record",
    "microvm_session_registry_model_definition",
    "microvm_session_registry_partition_key",
    "microvm_session_registry_sort_key",
    "microvm_session_registry_table_name",
    "microvm_session_token_metadata_from_provider_token",
    "validate_microvm_session_record",
    "validate_microvm_session_registry_record",
    "validate_microvm_session_status",
    "validate_microvm_session_token_metadata",
]
